File: mockito_fundamentals/src/test/java/com/techyourchance/mockitofundamentals/example7/LoginUseCaseSyncTest.java
package com.techyourchance.mockitofundamentals.example7;

import com.techyourchance.mockitofundamentals.example7.authtoken.AuthTokenCache;
import com.techyourchance.mockitofundamentals.example7.eventbus.EventBusPoster;
import com.techyourchance.mockitofundamentals.example7.eventbus.LoggedInEvent;
import com.techyourchance.mockitofundamentals.example7.networking.LoginHttpEndpointSync;
import com.techyourchance.mockitofundamentals.example7.networking.LoginHttpEndpointSync.EndpointResult;
import com.techyourchance.mockitofundamentals.example7.networking.LoginHttpEndpointSync.EndpointResultStatus;
import com.techyourchance.mockitofundamentals.example7.networking.NetworkErrorException;

import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;

import java.util.List;

import static org.hamcrest.CoreMatchers.instanceOf;
import static org.hamcrest.CoreMatchers.is;
import static org.junit.Assert.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class LoginUseCaseSyncTest {

    private static final String USERNAME = "Username";
    private static final String PASSWORD = "Password";
    private static final String AUTH_TOKEN = "authToken";
    private static final String NON_INITIALIZED_AUTH_TOKEN = "noAuthToken";

    // FIXME Change test double to mock
    LoginHttpEndpointSync mLoginHttpEndpointSyncMock;
    AuthTokenCache mAuthTokenCacheMock;
    EventBusPoster mEventBusPosterMock;

    LoginUseCaseSync SUT;

    @Before
    public void setup() throws Exception {
        mLoginHttpEndpointSyncMock = mock(LoginHttpEndpointSync.class);
        mAuthTokenCacheMock = mock(AuthTokenCache.class);
        mEventBusPosterMock = mock(EventBusPoster.class);
        SUT = new LoginUseCaseSync(mLoginHttpEndpointSyncMock, mAuthTokenCacheMock, mEventBusPosterMock);

        // given
        mockLoginSuccess();
    }

    @Test
    public void loginSync_success_usernameAndPasswordPassedToEndpoint() throws Exception {
        ArgumentCaptor<String> ac = ArgumentCaptor.forClass(String.class);

        // when
        SUT.loginSync(USERNAME, PASSWORD);

        // then
        verify(mLoginHttpEndpointSyncMock, times(1)).
                loginSync(ac.capture(), ac.capture());
        List<String> captures = ac.getAllValues();

        // then
        assertThat(captures.get(0), is(USERNAME));
        assertThat(captures.get(1), is(PASSWORD));
    }

    @Test
    public void loginSync_success_authTokenCached() throws Exception {
        ArgumentCaptor<String> ac = ArgumentCaptor.forClass(String.class);
        SUT.loginSync(USERNAME, PASSWORD);
        verify(mAuthTokenCacheMock).cacheAuthToken(ac.capture());
        assertThat(ac.getValue(), is(AUTH_TOKEN));
    }

    @Test
    public void loginSync_generalError_authTokenNotCached() throws Exception {
        mockGeneralError();
        SUT.loginSync(USERNAME, PASSWORD);
        verifyNoMoreInteractions(mAuthTokenCacheMock);
    }

    @Test
    public void loginSync_authError_authTokenNotCached() throws Exception {
        mockAuthError();
        SUT.loginSync(USERNAME, PASSWORD);
        verifyNoMoreInteractions(mAuthTokenCacheMock);
    }

    @Test
    public void loginSync_serverError_authTokenNotCached() throws Exception {
        mockServerError();
        SUT.loginSync(USERNAME, PASSWORD);
        verifyNoMoreInteractions(mAuthTokenCacheMock);
    }

    @Test
    public void loginSync_success_loggedInEventPosted() throws Exception {
        ArgumentCaptor<Object> ac = ArgumentCaptor.forClass(Object.class);
        SUT.loginSync(USERNAME, PASSWORD);
        verify(mEventBusPosterMock).postEvent(ac.capture());
        assertThat(ac.getValue(), is(instanceOf(LoggedInEvent.class)));
    }

    @Test
    public void loginSync_generalError_noInteractionWithEventBusPoster() throws Exception {
        mockGeneralError();
        SUT.loginSync(USERNAME, PASSWORD);
        verifyNoMoreInteractions(mEventBusPosterMock);
    }

    @Test
    public void loginSync_authError_noInteractionWithEventBusPoster() throws Exception {
        mockAuthError();
        SUT.loginSync(USERNAME, PASSWORD);
        verifyNoMoreInteractions(mEventBusPosterMock);
    }

    @Test
    public void loginSync_serverError_noInteractionWithEventBusPoster() throws Exception {
        mockServerError();
        SUT.loginSync(USERNAME, PASSWORD);
        verifyNoMoreInteractions(mEventBusPosterMock);
    }

    @Test
    public void loginSync_success_successReturned() throws Exception {
        LoginUseCaseSync.UseCaseResult result = SUT.loginSync(USERNAME, PASSWORD);
        assertThat(result, is(LoginUseCaseSync.UseCaseResult.SUCCESS));
    }

    @Test
    public void loginSync_serverError_failureReturned() throws Exception {
        mockServerError();
        LoginUseCaseSync.UseCaseResult result = SUT.loginSync(USERNAME, PASSWORD);
        assertThat(result, is(LoginUseCaseSync.UseCaseResult.FAILURE));
    }

    @Test
    public void loginSync_authError_failureReturned() throws Exception {
        mockAuthError();
        LoginUseCaseSync.UseCaseResult result = SUT.loginSync(USERNAME, PASSWORD);
        assertThat(result, is(LoginUseCaseSync.UseCaseResult.FAILURE));
    }

    @Test
    public void loginSync_generalError_failureReturned() throws Exception {
        mockGeneralError();
        LoginUseCaseSync.UseCaseResult result = SUT.loginSync(USERNAME, PASSWORD);
        assertThat(result, is(LoginUseCaseSync.UseCaseResult.FAILURE));
    }

    @Test
    public void loginSync_networkError_networkErrorReturned() throws Exception {
        mockNetworkError();
        LoginUseCaseSync.UseCaseResult result = SUT.loginSync(USERNAME, PASSWORD);
        assertThat(result, is(LoginUseCaseSync.UseCaseResult.NETWORK_ERROR));
    }

    private void mockLoginSuccess() throws NetworkErrorException {
        when(mLoginHttpEndpointSyncMock.loginSync(any(String.class), any(String.class)))
                .thenReturn(new EndpointResult(EndpointResultStatus.SUCCESS, AUTH_TOKEN));
    }

    private void mockNetworkError() throws Exception {
        doThrow(new NetworkErrorException())
                .when(mLoginHttpEndpointSyncMock).loginSync(any(String.class), any(String.class));
    }

    private void mockGeneralError() throws Exception {
        when(mLoginHttpEndpointSyncMock.loginSync(any(String.class), any(String.class)))
                .thenReturn(new LoginHttpEndpointSync.EndpointResult(
                        LoginHttpEndpointSync.EndpointResultStatus.GENERAL_ERROR, ""));
    }

    private void mockAuthError() throws Exception {
        when(mLoginHttpEndpointSyncMock.loginSync(any(String.class), any(String.class)))
                .thenReturn(new LoginHttpEndpointSync.EndpointResult(LoginHttpEndpointSync.EndpointResultStatus.AUTH_ERROR, ""));
    }

    private void mockServerError() throws Exception {
        when(mLoginHttpEndpointSyncMock.loginSync(any(String.class), any(String.class)))
                .thenReturn(new LoginHttpEndpointSync.EndpointResult(LoginHttpEndpointSync.EndpointResultStatus.SERVER_ERROR, ""));
    }
}

File: unit_testing_in_android/src/test/java/com/techyourchance/unittestinginandroid/myexample/LoginUseCaseSyncTest.kt
package com.techyourchance.unittestinginandroid.myexample

import com.nhaarman.mockitokotlin2.argumentCaptor
import com.nhaarman.mockitokotlin2.verify
import com.nhaarman.mockitokotlin2.whenever
import com.techyourchance.unittestinginandroid.myexample.authtoken.AuthTokenCache
import com.techyourchance.unittestinginandroid.myexample.eventbus.EventBusPoster
import com.techyourchance.unittestinginandroid.myexample.eventbus.LoggedInEvent
import com.techyourchance.unittestinginandroid.myexample.networking.LoginHttpEndpointSync
import com.techyourchance.unittestinginandroid.myexample.networking.LoginHttpEndpointSync.EndpointResult
import com.techyourchance.unittestinginandroid.myexample.networking.LoginHttpEndpointSync.EndpointResultStatus
import com.techyourchance.unittestinginandroid.myexample.networking.NetworkErrorException
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.instanceOf
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.MockitoJUnitRunner

private const val USERNAME = "username"
private const val PASSWORD = "password"
private const val AUTH_TOKEN = "authToken"
private const val NON_INITIALIZED_AUTH_TOKEN = "noAuthToken"

@RunWith(MockitoJUnitRunner::class)
class LoginUseCaseSyncTest {

	@Mock
	private lateinit var loginHttpEndpointSyncMock: LoginHttpEndpointSync

	@Mock
	private lateinit var authTokenCacheMock: AuthTokenCache

	@Mock
	private lateinit var eventBusPosterMock: EventBusPoster

	private lateinit var sut: LoginUseCaseSync

	@Before
	
	fun setup() {
		sut = LoginUseCaseSync(loginHttpEndpointSyncMock, authTokenCacheMock, eventBusPosterMock)
	}

	@Test
	fun loginSync_success_usernameAndPasswordPassedToEndpoint() {
		success()

		val ac = argumentCaptor<String>()

		sut.loginSync(USERNAME, PASSWORD)

		verify(loginHttpEndpointSyncMock).loginSync(ac.capture(), ac.capture())

		val id = ac.allValues[0]
		val pwd = ac.allValues[1]

		Assert.assertThat(id, `is`(USERNAME))
		Assert.assertThat(pwd, `is`(PASSWORD))
	}

	@Test
	fun loginSync_success_authTokenCached() {
		success()

		sut.loginSync(USERNAME, PASSWORD)

		verify(authTokenCacheMock).cacheAuthToken(anyString())
	}

	@Test
	fun loginSync_generalError_authTokenNotCached() {
		generalError()

		sut.loginSync(USERNAME, PASSWORD)

		verifyNoMoreInteractions(authTokenCacheMock)
	}

	@Test
	fun loginSync_authError_authTokenNotCached() {
		authError()

		sut.loginSync(USERNAME, PASSWORD)

		verifyNoMoreInteractions(authTokenCacheMock)
	}

	@Test
	fun loginSync_serverError_authTokenNotCached() {
		serverError()

		sut.loginSync(USERNAME, PASSWORD)

		verifyNoMoreInteractions(authTokenCacheMock)
	}

	@Test
	fun loginSync_success_loggedInEventPosted() {
		val ac = argumentCaptor<Any>()

		success()

		sut.loginSync(USERNAME, PASSWORD)

		verify(eventBusPosterMock).postEvent(ac.capture())

		val event = ac.allValues[0]

		assertThat(event, instanceOf(LoggedInEvent::class.java))
	}

	@Test
	fun loginSync_generalError_noInteractionWithEventBusPoster() {
		generalError()

		sut.loginSync(USERNAME, PASSWORD)

		verifyNoMoreInteractions(eventBusPosterMock)
	}

	@Test
	fun loginSync_authError_noInteractionWithEventBusPoster() {
		authError()

		sut.loginSync(USERNAME, PASSWORD)

		verifyNoMoreInteractions(eventBusPosterMock)
	}

	@Test
	fun loginSync_serverError_noInteractionWithEventBusPoster() {
		serverError()

		sut.loginSync(USERNAME, PASSWORD)

		verifyNoMoreInteractions(eventBusPosterMock)
	}

	@Test
	fun loginSync_success_successReturned() {
		success()

		val result = sut.loginSync(USERNAME, PASSWORD)

		assertThat(result, `is`(LoginUseCaseSync.UseCaseResult.SUCCESS))
	}

	@Test
	fun loginSync_serverError_failureReturned() {
		serverError()

		val result = sut.loginSync(USERNAME, PASSWORD)

		Assert.assertThat(result, `is`(LoginUseCaseSync.UseCaseResult.FAILURE))
	}

	@Test
	fun loginSync_authError_failureReturned() {
		authError()

		val result = sut.loginSync(USERNAME, PASSWORD)

		Assert.assertThat(result, `is`(LoginUseCaseSync.UseCaseResult.FAILURE))
	}

	@Test
	fun loginSync_generalError_failureReturned() {
		generalError()

		val result = sut.loginSync(USERNAME, PASSWORD)

		Assert.assertThat(result, `is`(LoginUseCaseSync.UseCaseResult.FAILURE))
	}

	@Test
	fun loginSync_networkError_networkErrorReturned() {
		networkError()

		val result = sut.loginSync(USERNAME, PASSWORD)

		Assert.assertThat(result, `is`(LoginUseCaseSync.UseCaseResult.NETWORK_ERROR))
	}

	private fun success() {
		whenever(loginHttpEndpointSyncMock.loginSync(anyString(), anyString()))
				.thenReturn(EndpointResult(EndpointResultStatus.SUCCESS, AUTH_TOKEN))
	}

	private fun generalError() {
		whenever(loginHttpEndpointSyncMock.loginSync(anyString(), anyString()))
				.thenReturn(EndpointResult(EndpointResultStatus.GENERAL_ERROR, NON_INITIALIZED_AUTH_TOKEN))
	}

	private fun authError() {
		whenever(loginHttpEndpointSyncMock.loginSync(anyString(), anyString()))
				.thenReturn(EndpointResult(EndpointResultStatus.GENERAL_ERROR, NON_INITIALIZED_AUTH_TOKEN))
	}

	private fun serverError() {
		whenever(loginHttpEndpointSyncMock.loginSync(anyString(), anyString()))
				.thenReturn(EndpointResult(EndpointResultStatus.SERVER_ERROR, NON_INITIALIZED_AUTH_TOKEN))
	}

	private fun networkError() {
		doThrow(NetworkErrorException())
				.whenever(loginHttpEndpointSyncMock).loginSync(anyString(), anyString())
	}
}
